### ISSUE TITLE
Setting Scope for Evaluation

### ISSUE THREAD

**Issue by nikeee** [uklimaschewski/EvalEx#156]
Is it possible to set a scope containing all variables for evaluation?
Something like this:
```Java
Map<String, BigDecimal> scope = new HashMap<>(); // contains values for a and b
scope.put("a", BigDecimal.ZERO);
scope.put("b", BigDecimal.ONE);
Expression ex = new Expression("a+b");
ex.eval(scope);
```
It would be nice if this wouldn't mutate the internal state of `ex`, so the expression can be used across multiple threads.

**Comment by RobertZenz** [comment]
What exactly are you trying to do, if I may ask? Using the same expression over multiple threads sounds interesting.

**Comment by nikeee** [comment]
I have a set of expressions which use variables and a lot of input values. I only want to parse and optimize the expression once. After that, I'd like to use the expression instance on multiple threads. Each thread uses the same set of variables, but with different values. This is only possible if the expression class is not mutable and therefore thread safe. Also, the set of variables is known and can be set all at once, instead of setting them one by one using `with`.

**Comment by RobertZenz** [comment]
I don't believe that a hard-fork would be necessary, instead the API could be extended to support this use-case. What would you think about something like this?

    Expression expression = new Expression("a+b+x");
    expression.with("a", BigDecimal.ZERO);
    
    Scope scope = new Scope();
    scope.with("b", BigDecimal.ZERO);
    scope.with("c", BigDecimal.ZERO);
    
    expression.eval(scope);

By default an `Expression` holds an "internal" scope of variables, but one can also set an additional scope for the evaluation. If a variable is defined in both, the external scope wins.

But note that I do not know whether `Expression` is actually thread-safe, that is a different topic I'm afraid.

**Comment by RobertZenz** [comment]
I don't believe that a hard-fork is necessary. The extension of the API which you outline in your first example seems easy enough. The only thing I'd change is that not a `Map` is accepted, but instead a domain class, more like this:

    Expression expression = new Expression("a+b+x");
    expression.with("a", BigDecimal.ZERO);
    
    Scope scope = new Scope();
    scope.with("b", BigDecimal.ZERO);
    scope.with("c", BigDecimal.ZERO);
    
    BigDecimal result = expression.eval(scope);

But I'm not sure whether `Expression` is thread-safe or not.

**Comment by nikeee** [comment]
This seems fine.

**Comment by RobertZenz** [comment]
@uklimaschewski I'd also like to take this one, though, I'm not sure in what timespan I can look at it. I'll do a pull request when I have something that I'm happy with so that you can double-check that it is something that you want. Of course I'll aim for zero API breakage.

**Comment by RobertZenz** [comment]
Yes, I have that in mind. But I'm not really familiar with the inner workings of `Expression`, so I have to test that first. I'm currently considering doing this in two steps, implementing the scope mechanism and getting everything thread-safe. I'm not sure if it will be in that order, but most likely.

**Comment by uklimaschewski** [comment]
Thank you for your effort, but to be honest, I think we are opening a can of ugly worms here.
Everything that is mutable has to go into the scope, including the MathContext, which you didn't have in the preview. And you would have to specify the scope for each operation then.
If somebody does an `expression.setPrecision(24).setRoundingMode(RoundingMode.DOWN).eval()`, which scope should it affect?
From my point of view, the expression instance is the scope. What is the advantage of introducing an inner scope, that holds all mutable members of the `Expression` class?
You wold have to create an instance of such a `Scope` in each thread anyway. So why not creating an instance of `Expression` in the first place? I see no benefit.

**Comment by nikeee** [comment]
That's what I feel like, too.

**Comment by RobertZenz** [comment]
Just to make it clear, I have no problem with you rejecting any of my changes (especially ones that I started working on without waiting for your reply first).

My thought behind this was that parsing the expression might be something expensive that one only wants to do once and then use that `Expression` instance to run on different values (even on different threads). That is my idea behind this, but I understand what you mean.

You are right, we *could* move all these properties into the `Scope`, but I wouldn't have done that. It is unlikely that you want to evaluate the same expression with different precisions, but what one might like to do is evaluate it with different values. Reusing `Expression`s would also become a lot more explicit (because currently one can only overwrite already added variables) and that the `Expression` can be used over multiple threads with ease is a nice feature to have, in my opinion, that is why I wanted to do this at all.

**Comment by RobertZenz** [comment]
I've thrown together the following test case:

~~~java
public class TestParsing {
	public static final void main(String[] args) {
		StringBuilder expressionStringBuilder = new StringBuilder("1+1");
		
		for (int counter = 0; counter < 1000; counter++) {
			expressionStringBuilder.append("+(5*7/(4*6*2)-4+1)");
		}
		
		String expressionString = expressionStringBuilder.toString();
		
		Expression expression = new Expression(expressionString);
		
		int count = 1000;
		
		long start = System.nanoTime();
		
		for (int counter = 0; counter < count; counter++) {
			expression.shuntingYard(expression.getExpression());
		}
		
		long end = System.nanoTime() - start;
		double duration = end / 1000l / 1000d;
		double durationEach = duration / (double)count;
		
		System.out.println("Total: " + Double.toString(duration) + "ms");
		System.out.println("Iteration: " + Double.toString(durationEach) + "ms");
	}
}
~~~

With `Expression.shuntingyard` being made `public`, and I get the following result:

    Total: 3550.056ms
    Iteration: 3.550056ms

So it seems like you are right, parsing the expression seems to be fast enough for most purposes.

I'll still explore this further because I'm intrigued by the idea. No pressure on your end. Maybe I'll end up with something that can be incorporated, if not, no problem.

**Comment by reachui** [comment]
Hi, I've started experimenting with EvalEx recently and it looks really promising for some basic arithmetic I'm looking at. I specifically searched out this 'Scope' question as it did occur to me that it would be very useful.

My use case is that I have several computations that I want to break into sections and use the results in subsequent computations.

public void doComputation(Map<String, BigDecimal> numbers)	{

	BigDecimal a = numbers.get("a");
	BigDecimal b = numbers.get("b");
	BigDecimal c = numbers.get("c");
	
	BigDecimal d = new Expression("(a + b) / c")
		.with("a", a)
		.with("b", b)
		.with("c", c)
		.eval();

	// I'd like to add "d" to the 'Scope' so that it can be used further on
	// i.e. I'm building up a list of varibles that are 'In Scope'

}

Hope this makes sense

**Comment by RobertZenz** [comment]
@reachui You mean something like this? (A little bit pseudo-code)

~~~java
Scope scope = new Scope();

scope.add("a", 5);
scope.add("b", 6);
scope.add("c", new Expression("a+b").eval(scope));
scope.add("d", new Expression("c*5+a").eval(scope));
scope.add("e", new Expression("d+c*a*b").eval(scope));
~~~

**Comment by reachui** [comment]
Yes, slightly different phrasing but the same thing essentially. Basically you're building the scope as you progress.

**Comment by RobertZenz** [comment]
As mentioned in #158, I'm currently not seeing how to implement this without breaking the API quite substantially, and for that the gains are a little, uh, little.

**Comment by nikeee** [comment]
I'd be ok with breaking API compability as long as it's released as a new major version.

**Comment by RobertZenz** [comment]
I've been thinking about doing that breakage from to time and also changing other parts (removing the inner classes, fixing that `Operator`s etc. don't have access to the current `MathContext`, static instances of the default operators/functions and so on), but I've been lacking the capacity so far to actually go through with it (after all, the API would need be changed very carefully). At the end of the day, there's also the question whether @uklimaschewski would accept such changes.

**Comment by nikeee** [comment]
My primary issue is that constructing the `Expression` does a Shunting-Yard and you cannot Re-use the expression with different values.

I played around with the library and came up with this:
https://github.com/uklimaschewski/EvalEx/compare/master...nikeee:master
Some notes on that:
- `EvaluationSettings`: things that matter only for evaluation (e. g. the MathContext) and not for parsing.
- `ExpressionSettings`: things that matter only for parsing (e. g. power operator precedence).
- Basic usage is something like this:
```
ExpressionLanguage l = new ExpressionLanguage();

ExpressionSettings s = new ExpressionSettings(1);
EvaluationSettings evalSettings = new EvaluationSettings(MathContext.DECIMAL32, true);

XXX expr = l.parseExpression("1+1", s);

System.out.println(expr.eval(evalSettings));
```

Where `XXX` is basically the `Expression` class mentioned in previous comments. Note that the previous `Expression` class was mostly refactored to `ExpressionLanguage` (which is basically the `Language` class mentioned in previous comments). This way, `Language.parseExpression` invokes shunting-yard only once. It's also immutable and can therefore be used in multiple threads with different scopes (which can be passed to an overload of the `eval` function).

With this refactoring, I can cut the time for a lot of computations down to 1/3 (previously, most of the time was spend in parsing/shunting the expression).

**Comment by RobertZenz** [comment]
That's an interesting approach. I like that this is taking the split even further, by making the keeping of functions and variables completely sparate. Though, I still belive in my naming approach of `Expression` and `Scope`, or maybe `Expression` and `Evaluation`. It would also be nice to provide a simplified instance for easier migration, one that works very similar to the current `Expression` and therefor limits he equired changes for an upgrade. I currently can't remember whether Udo was open to such a change, but I'd at least find it interesting to explore. Though, I'm a little bit unsure whether if I'm of much use right now.

**Comment by nikeee** [comment]
Yes, the current behavour could be replicated using an `Expression` class that uses these new APIs internally. This would make a migration pretty easy.

I also played around with separating the functions form the language. So we have 3 Things:
- Language: How to parse stuff as well as parsing itself
- Functions: What functions and constants are available (aka "environment")
- Expression: The thing that evaluates using a scope (aka "program")

In my sketch, language and the functions are mixed together. I decided against splitting this in my sketch because this would make porting the code much heavier and I wantet quick results. It also resembles a much larger breaking change. This could however, be mitigated by the above mentioned wrapper class.
